### PR TITLE
Exclude node at `rangeStart`

### LINKED ISSUE #8938: Range format should not format node near boundary
**Prettier 2.0.5**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzCEAEBeTBGAOgAYSiBuAIwEMAnHfY0yqgL20NPJABoQIAHGAEtoAZ2ShaNCAHcACrQTiUVADYyqAT3G8KNKmADWcGAGV+BoVADmyGDQCucXgAsYAW1UB1F0PiiLMDhTJT8hADc-TWRwUR0QK1E4Ghg5fWt3KmRUNSTeACtRAA8AIX0jE1MqdzgAGSs4bNznEEKi0ytrVTgARQcIeCbVPJALGiSaGOoKOFUeUZorGC8hABMYF2QADiJefmkkr31+GP24CfDG3gBHfvg0gWUQKlEAWig4OFWv+Zo4W6EfzSVAyWSQOWGLSS7iEdkcUM63T6A0a4OavBgVAoK3WmyQACYMfohKpOgBhCDuTIxc4AVnmDiSABUscoISNwk4AJJQb6wUxgRaCACCvNMME03SGI30NmCmJSyDwAGZeLLrHAAKK85D4gBsAF8DUA)
```sh
--parser babel
--range-end 26
--range-start 13
```

**Input:**
```jsx
foo = 1.0000;bar = 1.0000;baz=1.0000;
//           ^^^^^^^^^^^^^ Range
```

**Output:**
```jsx
foo = 1.0;
bar = 1.0;baz=1.0000;
```

**Expected behavior:**

Should only format `bar`, not `foo`

### PR DESCRIPTION
**Pull request by fisker**
Fixes #8938

Previously a node end at `rangeStart` in considered in range, now it's fixed.
And `rangeStart` and `rangeEnd` is not compared, because we treat them the same, no matter which one is smaller, but now we need know which one is the start, I decided to swap them if `rangeStart > rangeEnd`.

~Another issue was found before, I was going to fix it after this one, but it effect this PR, so I'm going to fix that one first, I will continue after it.~ Edit: Mistaken

- [x] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
This seems changes the range that only contains whitespace.

**Prettier 2.1.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzCEAUAdKACAwovXAQwBpcCAjXAShHJAgAcYBLaAZ2VFICd+EAO4AFAQh4pSAG2GkAnjybV+pMAGs4MAMot17KAHNkMfgFc4TOAFtqcACYPHAGVLHzpI3ABiEfjakMBzGyCCk5jAQjCAAFjA2MgDqsezwXPpgcDqSaewAbmkKYWBcyiCGXHD8MKJqRoHIqLJVTABWXAAeAEJqmto6pDZwLoZwTS1WIB2dOoZGMnAAiuYQ8BMyrSD6-FX8YdSk9jIxLPyGMEnsDjCxyAAcAAxMZxBVSWosYWdwe-njTAAjqt4HVWFJwlwALRQOCORwxfhwYHsJF1LyNJDNTZTKo2dimCy4+aLFZrcZYyZMGBHK43O5IABM1LU7Bk8wAwhAbJiQL8AKwxcxVAAqRyk2K2+UsAEkoM5YDowOc2ABBeU6GAKRYbLZqYzZGk1ZD8pj67wAUXlyAAjAA2AC+DqAA)
```sh
--parser babel
--range-end 16
--range-start 5
```

**Input:**
```jsx
foo(
           
a,
  b
)
```

**Output:**
```jsx
foo(a, b);
```

**Prettier pr-9704**
[Playground link](https://deploy-preview-9704--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAzCEAUAdKACAwovXAQwBpcCAjXAShHJAgAcYBLaAZ2VFICd+EAO4AFAQh4pSAG2GkAnjybV+pMAGs4MAMot17KAHNkMfgFc4TOAFtqcACYPHAGVLHzpI3ABiEfjakMBzGyCCk5jAQjCAAFjA2MgDqsezwXPpgcDqSaewAbmkKYWBcyiCGXHD8MKJqRoHIqLJVTABWXAAeAEJqmto6pDZwLoZwTS1WIB2dOoZGMnAAiuYQ8BMyrSD6-FX8YdSk9jIxLPyGMEnsDjCxyAAcAAxMZxBVSWosYWdwe-njTAAjqt4HVWFJwlwALRQOCORwxfhwYHsJF1LyNJDNTZTKo2dimCy4+aLFZrcZYyZMGBHK43O5IABM1LU7Bk8wAwhAbJiQL8AKwxcxVAAqRyk2K2+UsAEkoM5YDowOc2ABBeU6GAKRYbLZqYzZGk1ZD8pj67wAUXlyAAjAA2AC+DqAA)
```sh
--parser babel
--range-end 16
--range-start 5
```

**Input:**
```jsx
foo(
           
a,
  b
)
```

**Output:**
```jsx
foo(
           
a,
  b
)
```

**Comment by thorn0** [comment]
> This seems changes the range that only contains whitespace.

Can we avoid this? Looks unexpected.

**Comment by fisker** [comment]
We can find the source element around it, but what if

```js
a;<|>b;
```

Should we format `a` or `b`?

**Comment by fisker** [comment]
**Prettier 2.2.0**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEAhgBQCU+wAOlNSfqQNz7MsBGZD1A9FywHoC++AE6EoAczjUAviAA0ICAAdc0AM7JQhYcIgB3AAraEGlIQA2ewgE8NC1qLABrODADKSwmBwTkMYehwCnAAtqxwACYRkQAyYuLohJIAYhDCIYQwuL5m6DAQ8iAAFjAh5gDqRTjwap5gcG4m1TgAbtXWyOBqdiA+anDCMAai4hnIqBb9CgBWagAeAEKOLu6EIXAxPnDjk0Egs3NuPuLmcACK6BDwO+ZTIJ7C-cKdrITh5oVKwj4w5TgRMCKyAAHAAGBRfCD9cqiJSdL5wJ4tbYKACOl3gw2UphAhDUAFooHBIpFCsI4OicOThkkxkgJrc9v0Qjg-AEmcdThcrtt6bsFDA3n8AUCkAAmAWiHDmY4AYQgITpIERAFZCuh+gAVN6mBl3FqBACSUGisDcYG+KgAgia3DBrKcbndRBIGoLBsgxQBGBQuyQAURNnpV0mkQA)
```sh
--parser babel
--range-end 25
--range-start 21
```

**Input:**
```jsx
function a() {

a( );      b();
//   ^^^^ range
}
```

**Output:**
```jsx
function a() {

a();      b();
//   ^^^^ range
}
```

And this one, current logic will think range is `[21, 21]`, but if we can't find node at `22`, so format outer?

**Comment by thorn0** [comment]
In this case looks like it's okay not to format anything at all, but your previous example https://github.com/prettier/prettier/pull/9704#issuecomment-729397545 is different.

**Comment by fisker** [comment]
I updated this, 

- Ignore 0-width range, return originalText.
- "0-width range" didn't format anything, it doesn't make sense to format the "negative range", so if `rangeStart > rangeEnd` will skip format too.
- If all whitespace, there will be no difference as before, unless the `rangeStart` happened at `locEnd(node)`, because this is what this PR fixing. Internally, before we calculate as `[rangeStart, rangeStart]` (I think this unexpeced, it was not considered), now `[rangeEnd, rangeEnd]`.

**Comment by thorn0** [comment]
Is this expected?
**Prettier pr-9704**
[Playground link](https://deploy-preview-9704--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBqABAMwNzrxCACgB0o9yK9SBDAGlLwCNSBKEWkCABxgEtoAzslDUATqIgB3AApiEQlNQA2k6gE8hHRqOpgA1nBgBlLrt5QA5shiiArnA5wAtozgATN+4Ay1S7eoWcABiEKJO1DB8lsgg1LYwEOwgABYwTkoA6sm88AKmYHBG8jm8AG45ajFgApog5gJwojDSOhbhyJjKDRwAVgIAHgBCOvqGRtROcF7mcB1dDiB9-UbmFkpwAIq2EPBzSt0gpqINojGM1K5KSVyi5jAZvG4wycgAHAAMHDcQDRk6XDEbnATqVZhwAI7beAtbgKWICAC0UDg7ncSVEcEhvAxLQC7SQnX2CwaTl41jsxNW6y2O1mBPmHBgFweTxeSAATIydLwlKsAMIQJz4kDAgCsSVsDQAKhcFISDqV7ABJKCeWBGMC3HgAQVVRhganWewOOkshSZTWQABYOKbAgBRVXIADsAF9XUA)
```sh
--parser babel
--range-end 7
--range-start 4
```

**Input:**
```jsx
+ f;   oo(
           
a,
  b
)
```

**Output:**
```jsx
+ f;   oo(a, b);
```

**Comment by fisker** [comment]
I think yes, stable only format the text before, because the Node before rangeStart is found, now it's exlucded so range will be the whole text.

**Comment by fisker** [comment]
After a second thought, seems endNode include the node at `rangeEnd` too ??

File: tests/js/range/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -26,6 +26,78 @@ a = [, , , , , , , a];
...
+=====================================input======================================
+  1 | a (
+> 2 | );                 b (
+    |   ^^^^^^^^^^^^^^^^^
+  3 | );
+  4 | 
+=====================================output=====================================
+a (
+);                 b (
+);
```
Comment (fisker) [inline tests/js/range/__snapshots__/jsfmt.spec.js.snap:96]: I think I got it correct now. This one didn't get formatted is another issue I was going to fix after this one, actually the same as #9673.

**Review by thorn0 — APPROVED** [review]
LGTM now.
Looks like if `calculateRange` knew how to ignore comments, that would fix #9673.

**Pull request merged**